odoo/addons/base/models/res_users.py

```diff
@@ -997,20 +997,27 @@ def create(self, vals_list):
                 values['groups_id'] = type(self).groups_id.convert_to_write(gs, user)
         return super(UsersImplied, self).create(vals_list)
 
+    def _clear_demoted_user_access(self, values, users_before):
+        users_to_demote = users_before.filtered(
+            lambda u: not u.has_group('base.group_user')
+        )
+        if users_to_demote:
+            vals = {'groups_id': [(5, 0, 0)] + values['groups_id']}
+            super(UsersImplied, users_to_demote).write(vals)
+        return users_to_demote
+
     def write(self, values):
         users_before = self.filtered(lambda u: u.has_group('base.group_user'))
         res = super(UsersImplied, self).write(values)
         if values.get('groups_id'):
-            # add implied groups for all users
+            self._clear_demoted_user_access(values, users_before)
+            users_batch = defaultdict(self.browse)
```

Comment:
Add a comment: `# update users in batches, one batch per user groups`

Reply:
Isn't it obvious from code itself? I usually try to avoid comments that say the obvious. Variables like `users_batch`, `groups, `users` should describe what it does.